sky/serve/spot_policies.py

```diff
@@ -0,0 +1,123 @@
+""" Sky Spot Policy for SkyServe."""
+import collections
+import enum
+import typing
+from typing import Dict, List, Optional
+
+from sky import sky_logging
+
+if typing.TYPE_CHECKING:
+    from sky.serve import replica_managers
+    from sky.serve import service_spec
+
+logger = sky_logging.init_logger(__name__)
+
+
+class Location:
+    """Location class of a spot instance."""
+
+    def __init__(self, cloud: str, region: str, zone: Optional[str]) -> None:
+        self.cloud = cloud
+        self.region = region
+        self.zone = zone
+
+    def __eq__(self, other):
+        if isinstance(other, Location):
+            return (self.cloud == other.cloud and
+                    self.region == other.region and self.zone == other.zone)
+        return False
+
+    def __hash__(self):
+        return hash(self.cloud + self.region + self.zone)
```

Comment:
is it possible to add str and `None`? we need special handling in `self.zone` iiuc